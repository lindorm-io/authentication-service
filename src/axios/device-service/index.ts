export * from "./request-certificate-challenge";
export * from "./request-verify-certificate-challenge";
export * from "./request-verify-device-pin";
export * from "./request-verify-device-secret";
