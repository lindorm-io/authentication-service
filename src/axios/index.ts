export * from "./device-service";
export * from "./identity-service";
