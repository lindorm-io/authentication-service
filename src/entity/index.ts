export * from "./Account";
export * from "./RequestLimit";
export * from "./Session";
