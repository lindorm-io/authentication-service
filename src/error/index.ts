export * from "./AccountNotFoundError";
export * from "./AssertClientSecretError";
export * from "./AssertCodeChallengeError";
export * from "./AssertDeviceChallengeError";
export * from "./AssertOTPError";
export * from "./DeviceNotFoundError";
export * from "./ExpiredSessionError";
export * from "./InvalidAuthorizationError";
export * from "./InvalidAuthorizationTokenError";
export * from "./InvalidClientError";
export * from "./InvalidDeviceError";
export * from "./InvalidExpiryString";
export * from "./InvalidGrantTypeError";
export * from "./InvalidPermissionError";
export * from "./InvalidRefreshTokenError";
export * from "./InvalidScopeError";
export * from "./InvalidSubjectError";
export * from "./LockedAccountError";
export * from "./RejectedClientError";
export * from "./RequestLimitBackOffError";
