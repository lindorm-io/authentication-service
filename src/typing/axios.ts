import { ITokenIssuerSignData } from "@lindorm-io/jwt";

export interface IRequestChallengeVerifyData {
  challengeConfirmation: ITokenIssuerSignData;
}
