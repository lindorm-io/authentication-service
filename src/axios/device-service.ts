import axios from "axios";
import { Account, Session } from "../entity";
import { DEVICE_SERVICE_BASIC_AUTH, DEVICE_SERVICE_BASE_URL } from "../config";

export interface IVerifyDevicePINOptions {
  account: Account;
  deviceVerifier: string;
  pin: string;
  session: Session;
}

export const verifyDevicePIN = async (options: IVerifyDevicePINOptions): Promise<void> => {
  const url = new URL(`/headless/verify-pin`, DEVICE_SERVICE_BASE_URL);
  const { account, deviceVerifier, pin, session } = options;

  await axios.post(
    url.toString(),
    {
      account_id: account.id,
      device_challenge: session.authorization.deviceChallenge,
      device_id: session.deviceId,
      device_verifier: deviceVerifier,
      pin,
    },
    {
      auth: DEVICE_SERVICE_BASIC_AUTH,
    },
  );
};

export interface IVerifyDeviceSecretOptions {
  account: Account;
  deviceVerifier: string;
  secret: string;
  session: Session;
}

export const verifyDeviceSecret = async (options: IVerifyDeviceSecretOptions): Promise<void> => {
  const url = new URL(`/headless/verify-secret`, DEVICE_SERVICE_BASE_URL);
  const { account, deviceVerifier, secret, session } = options;

  await axios.post(
    url.toString(),
    {
      account_id: account.id,
      device_challenge: session.authorization.deviceChallenge,
      device_id: session.deviceId,
      device_verifier: deviceVerifier,
      secret,
    },
    {
      auth: DEVICE_SERVICE_BASIC_AUTH,
    },
  );
};
