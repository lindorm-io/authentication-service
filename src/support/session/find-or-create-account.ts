import { Account } from "../../entity";
import { IAuthContext } from "../../typing";
import { RepositoryEntityNotFoundError } from "@lindorm-io/mongo";
import { isLocked } from "@lindorm-io/jwt";
import { InvalidPermissionError } from "../../error";

export const findOrCreateAccount = (ctx: IAuthContext) => async (email: string): Promise<Account> => {
  const { repository } = ctx;

  try {
    const account = await repository.account.find({ email });

    if (isLocked(account.permission)) {
      throw new InvalidPermissionError();
    }

    return account;
  } catch (err) {
    if (err instanceof RepositoryEntityNotFoundError) {
      const account = new Account({ email });
      account.create();

      return repository.account.create(account);
    }

    throw err;
  }
};
