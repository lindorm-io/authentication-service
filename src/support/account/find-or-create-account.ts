import { Account } from "../../entity";
import { IKoaAuthContext } from "../../typing";
import { InvalidPermissionError } from "../../error";
import { ensureIdentity } from "../../axios";
import { isLocked } from "@lindorm-io/jwt";

export const findOrCreateAccount = (ctx: IKoaAuthContext) => async (email: string): Promise<Account> => {
  const { repository } = ctx;

  const account = await repository.account.findOrCreate({ email });

  if (isLocked(account.permission)) {
    throw new InvalidPermissionError();
  }

  await ensureIdentity(account.id);

  return account;
};
