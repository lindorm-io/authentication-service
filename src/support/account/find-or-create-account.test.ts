import MockDate from "mockdate";
import { Account } from "../../entity";
import { InvalidPermissionError } from "../../error";
import { Permission } from "@lindorm-io/jwt";
import { ensureIdentity } from "../../axios";
import { findOrCreateAccount } from "./find-or-create-account";
import { getGreyBoxRepository, resetStore } from "../../test";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "be3a62d1-24a0-401c-96dd-3aff95356811"),
}));
jest.mock("../../axios", () => ({
  ensureIdentity: jest.fn(() => ({})),
}));

MockDate.set("2020-01-01 08:00:00.000");

describe("findOrCreateAccount", () => {
  let ctx: any;
  let account: Account;

  beforeEach(async () => {
    ctx = {
      repository: await getGreyBoxRepository(),
    };
    account = await ctx.repository.account.create(
      new Account({
        email: "test@lindorm.io",
      }),
    );
  });

  afterEach(resetStore);

  test("should return an account", async () => {
    await expect(findOrCreateAccount(ctx)(account.email)).resolves.toMatchSnapshot();
    expect(ensureIdentity).toHaveBeenCalled();
  });

  test("should throw error when account is locked", async () => {
    account.permission = Permission.LOCKED;
    await ctx.repository.account.update(account);
    await expect(findOrCreateAccount(ctx)(account.email)).rejects.toStrictEqual(expect.any(InvalidPermissionError));
  });
});
