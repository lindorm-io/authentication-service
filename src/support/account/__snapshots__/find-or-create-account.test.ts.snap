// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findOrCreateAccount should return an account 1`] = `
Account {
  "_created": 2020-01-01T07:00:00.000Z,
  "_email": "test@lindorm.io",
  "_events": Array [
    Object {
      "date": 2020-01-01T07:00:00.000Z,
      "name": "account_created",
      "payload": Object {
        "created": 2020-01-01T07:00:00.000Z,
        "email": "test@lindorm.io",
        "otp": Object {
          "signature": null,
          "uri": null,
        },
        "password": Object {
          "signature": null,
          "updated": null,
        },
        "permission": "user",
        "updated": 2020-01-01T07:00:00.000Z,
      },
    },
  ],
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_identityId": null,
  "_otp": Object {
    "signature": null,
    "uri": null,
  },
  "_password": Object {
    "signature": null,
    "updated": null,
  },
  "_permission": "user",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;
